[inductor collectives] sink waits iterative

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157708
* #157706



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D77861763](https://our.internmc.facebook.com/intern/diff/D77861763)

IvanKobzarev committed:
[inductor collectives] sink waits iterative

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77861763).

IvanKobzarev committed:
Update on "[inductor collectives] sink waits iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77861763](https://our.internmc.facebook.com/intern/diff/D77861763)

[ghstack-poisoned]

wconstab commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -1613,6 +1614,7 @@ def _reorder_communication_preserving_peak_memory(
                 "reorder_for_compute_comm_overlap": True,
                 "reorder_for_compute_comm_overlap_passes": [
                     _reorder_communication_preserving_peak_memory,
```
i think it'd be good to add a test that shows that we can move a wait (wait_a) past a collective and past another wait (wait_b), in the case that wait_b and collective are not actually depending on the wait_a

IvanKobzarev committed:
Update on "[inductor collectives] sink waits iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77861763](https://our.internmc.facebook.com/intern/diff/D77861763)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77861763).

IvanKobzarev commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -1613,6 +1614,7 @@ def _reorder_communication_preserving_peak_memory(
                 "reorder_for_compute_comm_overlap": True,
                 "reorder_for_compute_comm_overlap_passes": [
                     _reorder_communication_preserving_peak_memory,
+                    sink_waits_iterative,
```
As we discussed offline - I've added a ban of swaps of groups that both contain collective_ops.
Once I added - the result of [reorder, sink_waits] and [sink_waits, reorder] became the same.

IvanKobzarev committed:
Update on "[inductor collectives] sink waits iterative"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77861763](https://our.internmc.facebook.com/intern/diff/D77861763)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77861763).
